src/backend/distributed/planner/distributed_planner.c

```diff
@@ -550,30 +562,120 @@ IsModifyDistributedPlan(DistributedPlan *distributedPlan)
 }
 
 
+/*
+ * PlanFastPathDistributedStmt creates a distributed planned statement using
+ * the FastPathPlanner.
+ */
+static PlannedStmt *
+PlanFastPathDistributedStmt(DistributedPlanningContext *ctx, Const *distributionKeyValue)
+{
+	ctx->plannerRestrictionContext->fastPathRestrictionContext->fastPathRouterQuery =
+		true;
+	ctx->plannerRestrictionContext->fastPathRestrictionContext->distributionKeyValue =
+		distributionKeyValue;
+
+	PlannedStmt *fastPathPlan = FastPathPlanner(ctx->originalQuery, ctx->parse,
+												ctx->boundParams);
+
+	return CreateDistributedPlannedStmt(fastPathPlan, ctx);
+}
+
+
+/*
+ * PlanDistributedStmt creates a distributed planned statement using the PG
+ * planner.
+ */
+static PlannedStmt *
+PlanDistributedStmt(DistributedPlanningContext *ctx,
+					List *rangeTableList,
+					int rteIdCounter)
+{
+	/*
+	 * Call into standard_planner because the Citus planner relies on both the
+	 * restriction information per table and parse tree transformations made by
+	 * postgres' planner.
+	 */
+	PlannedStmt *standardPlan = standard_planner(ctx->parse, ctx->cursorOptions,
+												 ctx->boundParams);
+
+	/* may've inlined new relation rtes */
+	rangeTableList = ExtractRangeTableEntryList(ctx->parse);
+	rteIdCounter = AssignRTEIdentities(rangeTableList, rteIdCounter);
+
+
+	PlannedStmt *result = CreateDistributedPlannedStmt(standardPlan, ctx);
+
+	bool setPartitionedTablesInherited = true;
+	AdjustPartitioningForDistributedPlanning(rangeTableList,
+											 setPartitionedTablesInherited);
+
+	return result;
+}
+
+
+/*
+ * PlanDelegatedFunctionCall creates a plan by delagating the function call to
+ * the worker.
+ */
+static PlannedStmt *
+PlanDelegatedFunctionCall(DistributedPlanningContext *ctx, DistributedPlan *delegatedPlan)
+{
+	PlannedStmt *result = standard_planner(ctx->parse, ctx->cursorOptions,
+										   ctx->boundParams);
+
+	result = FinalizePlan(result, delegatedPlan);
+
+	return result;
+}
+
+
+/*
+ * PlanNonDistributedStmt creates a normal (non-distributed) planned statement
+ * using the PG planner.
+ */
+static PlannedStmt *
+PlanNonDistributedStmt(DistributedPlanningContext *ctx, bool hasExternParam)
+{
+	PlannedStmt *result = standard_planner(ctx->parse, ctx->cursorOptions,
+										   ctx->boundParams);
+
+	if (hasExternParam)
+	{
+		/*
+		 * As in CreateDistributedPlannedStmt, try dissuade planner when planning
+		 * potentially failed due to unresolved prepared statement parameters.
+		 */
+		result->planTree->total_cost = FLT_MAX / 100000000;
```

Comment:
One of the reasons I wanted a separate `PlanDelegatedFunctionCall` is that this hasExternParam variable is misleading. It is only set to true deep down in `TryToDelegateFunctionCall` when we've already established the statement is a distributed function call, but couldn't delegate because of a parameter. Hence this block should move to `PlanDelegatedFunctionCall`.

There's similar logic for cases where we could not use the router planner because of a parameter in `CreateDistributedPlannedStmt`.

Reply:
I changed it differently now then we discussed before in person.
I've added a `plan` field to `DistributedPlanningContext` and I'm setting the total_cost now directly in the `TryToDelegateFunctionCall` function. I think this makes more sense, since `TryToDelegateFunctionCall` can fail for a variety of reasons and in all these cases it should simply fallback to non distributed execution of the function.